src/Groups/types.jl

```diff
@@ -198,6 +214,68 @@ struct AutomorphismGroup{T} <: GAPGroup
   end
 end
 
+################################################################################
+#
+#  Composite Groups
+#
+################################################################################
+
+
+"""
+    DirectProductOfGroups
+Either direct product of two or more groups of any type, or subgroup of a direct product of groups.
+"""
+struct DirectProductOfGroups <: GAPGroup
+  X::GapObj
+  L::Vector{<:GAPGroup}   # list of groups
+  n::Int64           # length(L)
+  Xfull::GapObj      # direct product of the GAP groups of L
+  isfull::Bool     # true if G is direct product of the groups of L, false if it is a proper subgroup
+
+
+  function DirectProductOfGroups(G::GapObj, L::Vector{<:GAPGroup}, n::Int64, Xf::GapObj, isf::Bool)
+    z = new(G,L,n,Xf,isf)
+    return z
+  end
+end
+
+
+"""
+    SemidirectProductOfGroups{S,T}
+Semidirect product of two groups of type `S` and `T` respectively, or subgroup of a semidirect product of groups.
+"""
+struct SemidirectProductOfGroups{S<:GAPGroup, T<:GAPGroup} <: GAPGroup 
+  X::GapObj
+  N::S              # normal subgroup
+  H::T              # group acting on N
+  f::GAPGroupHomomorphism{T,AutomorphismGroup{S}}
+  Xfull::GapObj         # full semidirect product: X is a subgroup of Xfull. 
+  isfull::Bool     # true if X==Xfull
+
+  function SemidirectProductOfGroups(G::GapObj, N::S, H::T, f, Xf::GapObj, isf::Bool) where S where T
+    z = new{S,T}(G,N,H,f,Xf,isf)
+    return z
+  end
+end
+
+"""
+    WreathProduct
+Wreath product of a group `G` and a group of permutations `H`.
+"""
+struct WreathProduct <: GAPGroup
+  X::GapObj
+  G::GAPGroup
+  H::PermGroup
+  Xfull::GapObj            # if H does not move all the points, this is the wreath product of (G, Sym(deg(H))
+  isfull::Bool             # true if Xfull == X
+
+  function WreathProduct(X::GapObj, G::GAPGroup, H::PermGroup, Xf::GapObj, isf::Bool)
+     z = new(X,G,H,Xf,isf)
+     return z
+  end
```

Comment:
Redundant

Reply:
But that's the same function created in the other types...

Reply:
It's redundant there, too. You only need to provide the "trivial" inner constructor (i.e. an inner constructor whose arguments correspond 1-1 to the members of the struct) if you also provide other non-trivial inner constructor, *and* really want the user to be able to access the "trivial" inner construct. Because if you don't provide any inner constructors, Julia automatically provides one.

Reply:
Yes, in this case it is useless. In some types (as `PermGroup`, `MatrixGroup`, etc.) it was created to be sure that the underlying GAP object has the correct type (indeed it contains a line like
> @assert GAP.Globals.IsPermGroup(...)

or similar. But it is not necessary for the user at the higher level.